### ISSUE TITLE
Add "variables" option to `no-use-before-define` rule

### ISSUE THREAD

**Issue by feross** [eslint/eslint#7111]
- **ESLint Version:** 3.5.0

I want `no-use-before-define` to check for classes or functions that are used before they are defined. But I do not want to check for variables that are used before they are defined.

Right now there exist two options, which modify whether functions and classes are checked. For example, I can disable these two checks:

```
"no-use-before-define": [2, { "functions": false, "classes": false }]
```

But I cannot disable checking if variables are used before they are defined. I propose adding a "variables" option to address this:

```
"no-use-before-define": [2, { "variables": false }]
```

There is already a precedent for an option like this, in the "blocks" option to the `padded-blocks` rule. It's used like this:

```
"padded-blocks": [2, { "blocks": "never", "switches": "never", "classes": "never" }]
```

**Comment by feross** [comment]
To make this clearer, here are some examples.

Examples of **incorrect** code for the `{ "variables": false }` option:

```js
/*eslint no-use-before-define: ["error", { "variables": false }]*/

console.log(foo);
var foo = 5;
```

Examples of **correct** code for the `{ "variables": false }` option:

```js
/*eslint no-use-before-define: ["error", { "variables": false }]*/

function foo() {
    console.log(foo);
}
var foo = 5;
```

The way this works would be 100% analogous to the "classes" option, but for variables. The existing "classes" option is documented here: https://github.com/eslint/eslint/blob/master/docs/rules/no-use-before-define.md#classes

**Comment by platinumazure** [comment]
I'm not opposed, but I'd love to see a real-world case where this is a problem (especially if it's something that "standard" recommends but cannot enforce, or wants to recommend but cannot, due to this limitation).

That said, since we already have classes and function exceptions, it does seem like we should just add a variable option. So one could make an argument that the [rule is incomplete](http://eslint.org/docs/developer-guide/contributing/rule-changes#accepting-a-rule-change) without this change.

Let's see what the rest of the team thinks.

**Comment by feross** [comment]
> I'm not opposed, but I'd love to see a real-world case where this is a problem (especially if it's something that "standard" recommends but cannot enforce, or wants to recommend but cannot, due to this limitation).

In `standard`, we would like to prevent clearly incorrect use-before-define cases like the following:

```js
console.log(foo);
var foo = 5;
```

While still allowing cases that may be correct if the program executes in the right order:

```js

function printFoo() {
     // Variable "foo" is not used before it is defined, even though it
     // appears in the source before foo is defined.
    console.log(foo);
}
var foo = 5;
printFoo();
```

There are just too many false positives for us when the latter example is disallowed.

**Comment by feross** [comment]
More specifically, here is the concrete code example that prompted this issue. We would like to disallow the following:

```js
var variable1 = variable1
let variable2 = variable2
const variable3 = variable3
```

**Comment by dcousens** [comment]
Another example that happens quite often and is impossible with this rule enabled:

``` js
function foo (x) {
  return x ? bar(x) : 1
}

function bar (x) {
  return foo(x - 1)
}
```

**Comment by not-an-aardvark** [comment]
This seems like a reasonable enhancement. I'll champion this proposal.

@eslint/eslint-team Any support for this proposal? I think [this comment](/eslint/eslint/issues/7111#issuecomment-246179695) explains the use-case well. We already have a `classes` option that does the same thing for classes.

**Comment by platinumazure** [comment]
Endorsed. Apologies to @feross for losing track of this and thank you @not-an-aardvark for picking this back up.

**Comment by feross** [comment]
No apology needed, @platinumazure! Thanks for the work both of you do on ESLint!

**Comment by not-an-aardvark** [comment]
@eslint/eslint-team Any support for this proposal?? We need 2 more :+1:s to accept it. This would allow users to avoid temporal-deadzone variable references without having to adhere to the stylistic requirements of `no-use-before-define`.

**Comment by dcousens** [comment]
@not-an-aardvark I added a +1.

**Comment by dcousens** [comment]
ping @fanatid @jprichardson @afk11

**Comment by not-an-aardvark** [comment]
@dcousens Thanks, I appreciate it. However, I was mainly referring to :+1:s from the ESLint team (from our [maintainer guide](http://eslint.org/docs/maintainer-guide/issues.html#consensus), we accept enhancements after they have a :+1: from three team members).

**Comment by dcousens** [comment]
@not-an-aardvark ah, sorry for the call out then everyone!

**Comment by fanatid** [comment]
@feross as I understand next code with `{ "variables": false }` will be **incorrect**?

```js
/*eslint no-use-before-define: ["error", { "variables": false }]*/

x()
function x () { console.log(42) }
```

**Comment by not-an-aardvark** [comment]
@fanatid That's correct (the code there is already incorrect with this rule, and the `variables` option would only affect variable declarations, not function declarations).

**Comment by feross** [comment]
@not-an-aardvark Are you sure? @fanatid's example is a function so it shouldn't not affected by `{ "variables": false }`. Also, with `{ "functions": false }` the code is indeed valid.

This is not currently an error:

```js
/* eslint no-use-before-define: ["error", { "functions": false }] */

x()
function x () { console.log(42) }
```

**Comment by not-an-aardvark** [comment]
@feross I'm not sure I understood what you're saying, but I think we're in agreement.

This should report an error:

```js
/* eslint no-use-before-define: ["error", { "variables": false }] */

x()
function x () { console.log(42) }
```

This should not report an error:

```js
/* eslint no-use-before-define: ["error", { "functions": false }] */

x()
function x () { console.log(42) }
```

**Comment by feross** [comment]
Yes, sorry for the confusing wording. We're in agreement.

**Comment by Kovensky** [comment]
This rule, as it is, also has false positives with binding `export`s:

```js
export { group as default }
const group = new Map()
```

**Comment by not-an-aardvark** [comment]
@Kovensky, may you please create another issue for that? If there is a false positive with the rule then we certainly want to address it, but it seems unrelated to this proposal.

**Comment by Kovensky** [comment]
@not-an-aardvark hm, true, it is *kinda* related but not exactly.

At any rate, have there been any updates w.r.t. deciding on whether to implement the potential `variables` option in code? It'd be nice to have a static TDZ checker without enforcing style changes.

Just noticed this after intentionally putting a `let` in TDZ for most of a file to make its initialization point clearer...

**Comment by Kovensky** [comment]
Opened https://github.com/eslint/eslint/issues/7858

**Comment by feross** [comment]
What's the status of this issue? Are we still waiting for 👍's from additional ESLint team members? Thanks.

**Comment by vitorbal** [comment]
I'll support this. @eslint/eslint-team looks like all we need is one more +1 to accept this issue (and we already have a champion). Any takers?

**Comment by mikesherov** [comment]
👍

**Comment by not-an-aardvark** [comment]
PR up at https://github.com/eslint/eslint/issues/7948
